packages/angular/src/generators/ng-add/utilities/file-change-recorder.ts

```diff
@@ -0,0 +1,44 @@
+import type { Tree } from '@nrwl/devkit';
+import MagicString from 'magic-string';
```

Comment:
I'm not sure that I like introducing the dependency on this package. It seems like the biggest benefit from it would be generating source maps, and we don't use that. Could this just be a set of simple methods?

I'd argue we don't even really need a "mutable" content, rather appendLeft could easily look like:

```typescript
this.currentContents = this.currentContents.substring(0, index) + insertedString + this.currentContents.substring(index)
```

I'm not really sure that its preferable to change this out, its just something to think about.

Reply:
I'll have to defer to @leosvelperez to answer this as this is a part of his work for the angular cli migration to support cypress 10 and not super sure on what his requirements were for this.

Reply:
@AgentEnder this is not just doing simple string transformations like appendLeft. What `magic-string` allows you to do is to work with a string as if it hasn't changed, so you can keep transforming it based on the original contents and the positions you might have found with the original content. So, that allows us to keep working with the same sourceFile and nodes, using their positions based on the original string, and still be able to continue changing the string without being affected by the positions shifting because the content changes. If we were to do a simple append like the one you suggested, we'd have to keep updating the sourceFile after each modification and re-querying stuff to handle the updated content.

This is pretty handy when dealing with certain AST transformations, like the ones in this PR. The `magic-string` package is used by the Angular DevKit for this same purpose.

Reply:
Could you use `applyChangesToString` for this?

https://github.com/nrwl/nx/blob/d007d37fb4f625fc4854d06d2e083ed778d6a3db/packages/devkit/src/utils/string-change.spec.ts